tls/s2n_server_new_session_ticket.c

```diff
@@ -253,3 +253,76 @@ int s2n_tls13_server_nst_write(struct s2n_connection *conn, struct s2n_stuffer *
 
     return S2N_SUCCESS;
 }
+
+/**
+ *= https://tools.ietf.org/rfc/rfc8446#section-4.6.1
+ *#     struct {
+ *#         uint32 ticket_lifetime;
+ *#         uint32 ticket_age_add;
+ *#         opaque ticket_nonce<0..255>;
+ *#         opaque ticket<1..2^16-1>;
+ *#         Extension extensions<0..2^16-2>;
+ *#      } NewSessionTicket;
+**/
+int s2n_tls13_server_nst_recv(struct s2n_connection *conn, struct s2n_stuffer *input)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(input);
+
+    POSIX_ENSURE(conn->mode == S2N_CLIENT && conn->actual_protocol_version >= S2N_TLS13, S2N_ERR_SAFETY);
+
+    if (conn->config->session_ticket_cb != NULL) {
+        uint32_t ticket_lifetime = 0;
+        POSIX_GUARD(s2n_stuffer_read_uint32(input, &ticket_lifetime));
+        /**
+         *= https://tools.ietf.org/rfc/rfc8446#section-4.6.1
+         *# The value of zero indicates that the
+         *# ticket should be discarded immediately.
+         */
+        if (ticket_lifetime == 0) {
+            return S2N_SUCCESS;
+        }
+        struct s2n_ticket_fields ticket_fields = { 0 };
+        POSIX_GUARD(s2n_stuffer_read_uint32(input, &ticket_fields.ticket_age_add));
+
+        uint8_t ticket_nonce_len = 0;
+        struct s2n_blob nonce = { 0 };
+        POSIX_GUARD(s2n_stuffer_read_uint8(input, &ticket_nonce_len));
+        uint8_t nonce_data[UINT8_MAX] = { 0 };
+        POSIX_GUARD(s2n_blob_init(&nonce, nonce_data, ticket_nonce_len));
+        POSIX_GUARD(s2n_stuffer_read_bytes(input, nonce.data, ticket_nonce_len));
+
+        uint16_t session_ticket_len = 0;
+        POSIX_GUARD(s2n_stuffer_read_uint16(input, &session_ticket_len));
+        POSIX_ENSURE(session_ticket_len > 0, S2N_ERR_SAFETY);
+        POSIX_GUARD(s2n_realloc(&conn->client_ticket, session_ticket_len));
+        POSIX_GUARD(s2n_stuffer_read(input, &conn->client_ticket));
+
+        /* Derive individual session ticket secret */
+        s2n_tls13_connection_keys(secrets, conn);
+        struct s2n_blob master_secret = { 0 };
+        POSIX_GUARD(s2n_blob_init(&master_secret, conn->resumption_master_secret, secrets.size));
+        uint8_t session_secret_data[S2N_TLS13_SECRET_MAX_LEN] = { 0 };
+        POSIX_GUARD(s2n_blob_init(&ticket_fields.session_secret, session_secret_data, secrets.size));
+        POSIX_GUARD_RESULT(s2n_tls13_derive_session_ticket_secret(&secrets, &master_secret, &nonce, &ticket_fields.session_secret));
```

Comment:
Since this calculation is done both when sending and receiving, is it worth moving it to a separate method shared by both? Might make the testing easier

Reply:
The issue is, if I move secrets to a new function, then I don't know how big the secret blob should be for ticket_fields.session_secret (see line 306.) And I'm pretty sure that when derive the secret, that blob size needs to be accurate.